media/multipart/src/main/java/io/helidon/media/multipart/MultiPartDecoder.java

```diff
@@ -247,85 +392,148 @@ private BodyPartChunk createPartChunk(BufferEntry entry) {
+                    // entire bufferEntryIterator, and signal onError
+                    long curr = n <= 0
+                                ? chunksRequested.getAndSet(-1)
+                                : chunksRequested.getAndUpdate(v -> Long.MAX_VALUE - v > n
+                                ? v + n : v < 0 ? v == Long.MIN_VALUE ? n : v : Long.MAX_VALUE);
+                    if (curr == 0) {
+                        MultiPartDecoder.this.drain();
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                    cancelled = true;
+                    // Ensure the part chunks are drained to make the next part available
+                    if (chunksRequested.getAndSet(-1) == 0) {
+                        MultiPartDecoder.this.drain();
+                    }
+                }
+            });
+
+            if (chunksRequested.compareAndSet(Long.MIN_VALUE, 0)) {
+                return;
             }
-            lastEvent = event;
+            MultiPartDecoder.this.drain();
         }
 
         /**
-         * Indicate if the parser has received any data.
-         *
-         * @return {@code true} if the parser has been offered data,
-         * {@code false} otherwise
+         * Set the next buffer entry iterator.
+         * @param iterator the iterator to set
          */
-        boolean isStarted() {
-            return lastEvent != null;
+        void nextIterator(Iterator<BufferEntry> iterator) {
+            // This is invoked only when the previous bufferEntryIterator has been consumed fully,
+            // and chunksRequested > 0, so no one is calling drain() concurrently
+            // chunksRequested is modified atomically, so any future invocation of drain() will observe
+            // bufferEntryIterator normal store (bufferEntryIterator and all of its content is published safely)
+            bufferEntryIterator = iterator;
         }
 
         /**
-         * Indicate if the parser has reached the end of the message.
+         * Complete the publisher.
          *
-         * @return {@code true} if completed, {@code false} otherwise
+         * @param th throwable, if not {@code null} signals {@code onError}, otherwise signals {@code onComplete}
          */
-        boolean isCompleted() {
-            return lastEvent.type() == MimeParser.EventType.END_MESSAGE;
+        void complete(Throwable th) {
+            if (cancelled) {
+                subscriber = null;
+                return;
+            }
+            cancelled = true;
+
+            // bufferEntryIterator is drained because complete() is invoked only by drain() which proceeds past
+            // state == BODY only when drain() returned true
+            if (th != null) {
+                subscriber.onError(th);
+            } else if (chunksRequested.get() < 0) {
+                subscriber.onError(new IllegalArgumentException("Expecting only positive requests"));
```

Comment:
Move to line `471` and change the condition to:

```
if (chunksRequested.get() < 0 && !cancelled && th == null) {
   th = new IllegalArgumentException("Expecting only positive requests for content");
}
```

(Because currently a `cancel` concurrent with `complete` will not be distinguishable from bad request)